lib/foreman_rh_cloud/engine.rb

```diff
@@ -210,6 +210,31 @@ def self.register_scheduled_task(task_class, cronline)
       )
     end
 
+    # Ideally this code belongs to an initializer. The problem is that Katello controllers are not initialized completely until after the end of the to_prepare blocks
+    # This means I can patch the controller only in the after_initialize block that is promised to run after the to_prepare
+    # initializer 'foreman_rh_cloud.allow_smart_proxy_actions', :before => :finisher_hook, :after => 'katello.register_plugin'  do |_app|
+    # end
+    config.after_initialize do
+      # skip overrides in migrations, since the controller initialization depends on tables existense
+      if defined?(Katello) && ActiveRecord::Base.connection.data_source_exists?(SmartProxy.table_name)
```

Comment:
Perhaps this is cleaner:
```suggestion
      if defined?(Katello) && !Foreman.in_setup_db_rake?
```

It specifically checks for the setup tasks:
https://github.com/theforeman/foreman/blob/f4246cdaca1d0039d93de65cda9e52862c02f81b/lib/foreman.rb#L25-L27

Also, isn't Katello a hard dependency? It has `require 'katello'` as the first line so I'd think `defined?(Katello)` is always true.

Though I'm also wondering why this already needs DB access. What actually loads the DB? That's a bug in itself.

Reply:
> Though I'm also wondering why this already needs DB access. What actually loads the DB? That's a bug in itself.

I am getting the following call stack if I disable the check:

```
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "smart_proxies" does not exist
...
/home/vagrant/foreman/app/models/smart_proxy.rb:44:in `block in <class:SmartProxy>'
...
/home/vagrant/katello/app/models/katello/concerns/smart_proxy_extensions.rb:88:in `pulp_primary'
/home/vagrant/katello/app/services/katello/repository_type_manager.rb:13:in `pulp_primary'
/home/vagrant/katello/app/services/katello/repository_type_manager.rb:34:in `enabled_repository_types'
/home/vagrant/katello/app/services/katello/repository_type_manager.rb:53:in `generic_repository_types'
/home/vagrant/katello/app/controllers/katello/api/v2/repositories_controller.rb:505:in `<class:RepositoriesController>'
```
If I understand the error correctly, the controller assumes that the database is already created and makes some DB queries directly from the module code. I think the proper solution will work around it, since the checks will be dynamic.

Reply:
> If I understand the error correctly, the controller assumes that the database is already created and makes some DB queries directly from the module code. I think the proper solution will work around it, since the checks will be dynamic.

Hmm, I think we should make that code safe.

For context, it's https://github.com/Katello/katello/blob/35895ec09cbeaaa7881569c8d3a0ef526e3bb21d/app/controllers/katello/api/v2/repositories_controller.rb#L505-L510

The background is that the supported repository types is dynamically determined based on the content types Pulp supports. Those are stored as capabilities on the Smart Proxy with the Pulp feature.

@ofedoren do you have any suggestions on how to improve that pattern? Effectively make the params lazy loaded.

Reply:
I have to admit I don't like the fact that I reference the controllers so early in the process. As I have said, in the proper solution, we won't need to load the controller class at all.
Ideally I would like to mark something like this:
``` ruby
plugin do
  enable_proxy_auth_for controller: 'foo', action: 'bar'
``

Reply:
> @ofedoren do you have any suggestions on how to improve that pattern? Effectively make the params lazy loaded.

As of now I don't have any. I'm afraid we can't do much without introducing something new in apipie-rails lib. I'll play around with one rather nasty idea to see if it helps.

Reply:
@ekohl, I think I've found a slight but bearable workaround: https://github.com/Katello/katello/pull/11430

UPD: I didn't find anything, a patch in `apipie-rails` is pending :/